Fallible APIs for StringInterner for #70

royaltm opened the pull request:
Hi @Robbepop!

This is my take on issue #70.

This pull request is __NOT READY__ to merge.
I haven't add any tests nor did I check whether new code breaks anything in the existing test suites.

Let's iterate this through here instead.

royaltm commented:
First, I haven't found issues adding fallible API to Backend implementations (except the capacity issue, below).

The most problematic change involves the `dedup` HashMap:
- to implement `try_reserve` I needed to directly access HashTable, thus a `raw` feature needed to be enabled for `hashbrown`
- there's a substanstial addition here: https://github.com/royaltm/string-interner/blob/e8a7616160f8834f53927d72532aab12ec27bf2a/src/interner.rs#L163
 which might beg for a bigger refactor to this function

I'm also a bit concerned about the sanity of the `with_capacity` and the new `try_reserve` API, since the implementation relies on guesswork - estimating average word size by google - like with everything "on average" will mostly be of no use for most particular cases.

What if instead we give a user a way to estimate the "average" string size? Or perhaps even the separate additional capacity for string data.

Robbepop commented:
@royaltm thank you for the PR!

I think this PR generally goes into the right direction.

The `RawMap` API should have been implemented and used by the `StringInterner` long ago, but I didn't get to it. Maybe it would be nice to separate this change into another PR and basing this PR on it.

CI is currently unhappy with docs and formatting. Also a memory consumption test is broken because the number of expected allocations decreased. Can you explain why?

I am a bit sceptic about the usefulness of the reserve API for `StringInterner` in general after seeing the mess that it introduces. The `reserve` API was once useful when we still had the `SimpleBackend`. However, the `SimpleBackend` got removed because it was too inefficient and thus served no real purpose. Without it the `reserve` API depends too much on heuristics and should probably be removed or redesigned to not be based on the expected number of strings to intern but on the expected bytes of all expected strings to be interned but even then it wouldn't be ideal imo.

royaltm commented:
Perhaps different number of allocations must have something to do with changes to the buffer backend where I introduced the `calculate_var7_size` function so I am able to reserve space for the next entry including the var7 size marker ahead of pushing to the backend buffer.
Now the buffer backend is resized only once before `encode_var_usize` is called and the string itself is pushed.
Previously var7 size marker was pushed separately from the string itself and each operation could possibly reallocate backend buffer. But that's just my guess.

As for the RawTable API, I'm not quite sure what advantage it would have over current implementation. Perhaps just less intermediary function calls?

Robbepop commented:
> As for the RawTable API, I'm not quite sure what advantage it would have over current implementation. Perhaps just less intermediary function calls?

I have read somewhere that the `RawEntry` API of the `hashbrown` crate was to be phased out in favor of the newer and more versatile `RawMap` type. Since you are now making use of the `raw` crate feature I thought it would make sense to go full in with the change. Not necessary though. We can go ahead without it.

royaltm commented:
I suppose moving to `RawMap` interface should be only an internal change. I propose we can tackle this issue *after* the one at hand.

Regarding `try_reserve` and `with_capacity` interface I agree that its usefulness is pretty questionable. One thing is that It's really hard to make predictions ahead. Especially with the `BucketBackend` and what makes the problem even worse is that we can't `shrink_to_fit` the head String due to the known problems with reallocation.

Perhaps I'll remove the `try_reserve` calls from API or make them internal where they are still required, and we'll see where that get us.

Robbepop commented:
> I suppose moving to `RawMap` interface should be only an internal change. I propose we can tackle this issue _after_ the one at hand.
> 
> Regarding `try_reserve` and `with_capacity` interface I agree that its usefulness is pretty questionable. One thing is that It's really hard to make predictions ahead. Especially with the `BucketBackend` and what makes the problem even worse is that we can't `shrink_to_fit` the head String due to the known problems with reallocation.
> 
> Perhaps I'll remove the `try_reserve` calls from API or make them internal where they are still required, and we'll see where that get us.

The `try_reserve` probably does not make sense for the `Backend` to implement since it is too backend specific. At the moment no backend has a proper and well defined way to do it. For the `StringInterner` itself `try_reserve` could still be useful for its hashmap to deduplicate strings, however, it is probably not much help in practice.

royaltm commented:
Hi @Robbepop!

Sorry for the delay, I've had a whirlwind of obligations to fulfill in the passing weeks and I'd rather not contribute at all than contribute haphazardly during my lunch break as good and quickly seldom meet.

* I removed the `Backend::try_reserve` from trait and backend implementations,
* I left `StringInterner::try_reserve` with a doc note about what it does exactly,
* fixed docs issues,
* added a mention of `try_get_or_intern` in the root doc of the `StringInterner`.
* testing of `calculate_var7_size` together with each `encode_var_usize`,
* updated number of (de)allocations in `buffer_backend::test_memory_consumption`.

If this meets your criteria I don't think there's anything left considering the scope of this PR.

royaltm commented:
Hi @Robbepop,

I've fixed formatting errors, I'm not sure however what can I do to fix the tarpaulin issue.
Is there anything else on my end that needs to be done?

Robbepop commented:
I just ran benchmarks for this branch locally and compared them to the current main branch.

Unfortunately I saw quite some big performance regressions with this PR:

- `get_or_intern/fill-empty/new/`: ~10% regression
- `get_or_intern/fill-empty/with_capacity/`: ~20-30% regression
- `get_or_intern/already-filled/`: ~5% regression
- `get_or_intern_static`: `BuckedBackend` improved by 5%, other backends regressed by 10%

Robbepop commented on src/backend/buffer.rs:
```diff
@@ -123,30 +123,39 @@ where
-    fn encode_var_usize(&mut self, value: usize) -> usize {
-        encode_var_usize(&mut self.buffer, value)
+    fn try_encode_var_length(&mut self, length: usize) -> Result<()> {
+        let add_len = length + calculate_var7_size(length);
```
I would rename it to something like `len_encode_var_usize` which is very similar to `encode_var_usize` to draw the parallels to the reader of the code.

Robbepop replied:
Though I have to admit that it is a bit unfortunate that this encoding step is now 2 phased.

royaltm replied:
`encode_var_usize` pushes each byte separately, calling `try_reserve` every time a byte is encoded would take down the performance even more. However with `push_within_capacity` stabilized that could be implemented differently.

Alternative solution is to encode length into an array on the stack and then call `extend_from_slice` on `self.buffer`. This however can't be done efficiently without a bit of unsafe code (`MaybeUninit`) or would require using ArrayVec or a similar crate.

Robbepop commented on src/backend/buffer.rs:
```diff
@@ -123,30 +123,39 @@ where
     }
 }
 
+/// According to google the approx. word length is 5.
+const DEFAULT_STR_LEN: usize = 5;
```
We should really get rid of this thing entirely. My bad to have introduced them in the first place. 🙈

royaltm replied:
By that, do you mean to remove `with_capacity` from the Backend altogether?

Robbepop commented on src/backend/buffer.rs:
```diff
@@ -158,11 +167,9 @@ where
 
     #[cfg_attr(feature = "inline-more", inline)]
     fn with_capacity(capacity: usize) -> Self {
-        /// We encode the `usize` string length into the buffer as well.
-        const LEN_USIZE: usize = mem::size_of::<usize>();
-        /// According to google the approx. word length is 5.
-        const DEFAULT_STR_LEN: usize = 5;
-        let bytes_per_string = DEFAULT_STR_LEN + LEN_USIZE;
+        // We encode the `usize` string length into the buffer as well.
+        let var7_len: usize = calculate_var7_size(capacity);
+        let bytes_per_string = DEFAULT_STR_LEN + var7_len;
```
As said in a prior comment, we should probably get rid of all the heuristics based `with_capacity` methods and get this down to the basics. A Rust `BTreeMap` for example also does not provide a `with_capacity` method because it simply does not make sense to do so.

royaltm replied:
I agree, it's rather pointless.

Robbepop commented on src/backend/buffer.rs:
```diff
@@ -200,6 +207,16 @@ where
     }
 }
 
+/// Calculate var7 encoded size from a given value.
+#[inline]
+fn calculate_var7_size(value: usize) -> usize {
+    // number of bits to encode
+    // value = 0 would give 0 bits, hence: |1, could be anything up to |0x7F as well
+    let bits = usize::BITS - (value | 1).leading_zeros();
+    // (bits to encode / 7).ceil()
+    ((bits + 6) / 7) as usize
+}
```
Can you please explain this formula a bit better so that it does not look like magic to the reader?

Robbepop replied:
Also: is this precise or an over approximation?

royaltm replied:
This formula is precise and this is also reflected in tests.

royaltm replied:
The formula goes like this:

1. Take the binary size (bit length) of the original length

```
let bits = usize::BITS - (value | 1).leading_zeros();
```
For example if your length is `42` this is binary `101010`, the bit length is 6.
Assuming `usize::BITS` is 64 `leading_zeros` of 42usize is 58. 64 - 58 = 6.

2. If the binary size is 0, make it 1 (0 is the only case that breakes the formula, hence `(value | 1)`).

3. The encoded size in bytes is the total number of encoded bits divided by 7 (each byte encodes only 7 bits of the original value), rounded up because the last byte's 7-bits might not be filled entirely: hence (bits + 6).

```
// (bits to encode / 7).ceil()
((bits + 6) / 7) as usize
```

For example 0x7F bit length is 7, (7 + 6) / 7 == 1

The length is encoded as: 0b0111_1111

0x80 bit length is 8, (8 + 6) / 7 == 2

The length is encoded as: 0b1000_0000 0b0000_0001

royaltm replied:
Let's take the original bit value of the encoded length (each bit placement is represented by a character)

ZYXWVUTSRQPONMLKJIHGFEDCBA9876543210 (bit-length: 36)

The encoded chunks end up being:

!6543210 !DCBA987 !KJIHGFE !RQPONML !YXWVUTS _______Z

Where `!` and `_` are `1` and `0` bit markers respectively.

As seen clearly each 7 bits ends up in each chunk with the last chunk filled with the remaining 1-7 bits

(36 + 6) / 7 == 6

Now if we take:

YXWVUTSRQPONMLKJIHGFEDCBA9876543210 (bit-length: 35)

The encoded chunks end up being:

!6543210 !DCBA987 !KJIHGFE !RQPONML _YXWVUTS

(35 + 6) / 7 == 5

This process repeats every 7 bits.

The only exception is the `0` itself which if we followed the formula precisely should be encoded in exactly `0` bytes - because there are no bits to encode!

This leads us to the interesting observation that we could in fact extend the encoded var7 by adding any number of redundant `0` at the end, and the decoding formula (providing we have an infinite integer capacity) would still work:

!6543210 !DCBA987 !KJIHGFE !RQPONML !YXWVUTS ________
!6543210 !DCBA987 !KJIHGFE !RQPONML !YXWVUTS !_______ ________
!6543210 !DCBA987 !KJIHGFE !RQPONML !YXWVUTS !_______ !_______ ________

and so on...

I hope that explains why `0` is breaking the formula without actually really breaking it. What we really need is just 1 bit of information that this is the end of the stream of data.

Robbepop commented on src/backend/string.rs:
```diff
@@ -130,20 +131,35 @@ where
     }
 }
 
+// According to google the approx. word length is 5.
+const DEFAULT_WORD_LEN: usize = 5;
```
Again: we should remove these heuristics entirely.

Robbepop commented on src/interner.rs:
```diff
@@ -204,6 +224,13 @@ where
             backend,
         } = self;
         let hash = make_hash(hasher, string.as_ref());
+        // this call requires hashbrown `raw` feature enabled
```
please remove superflous comment

royaltm replied:
Sure, this was more of a message to you than the real code comment.

Robbepop commented:
All in all this looks good. The benchmark regressions need to be fixed before we can merge though.

royaltm commented on src/backend/bucket/mod.rs:
```diff
@@ -91,25 +91,25 @@ where
         Self {
             spans: Vec::with_capacity(cap),
-            head: FixedString::with_capacity(cap),
+            head: FixedString::try_with_capacity(cap).unwrap(),
```
Here I was actually assuming the error message from try_with_capacity will do the job.

royaltm commented on src/backend/bucket/mod.rs:
```diff
@@ -142,29 +142,54 @@ where
+        // FIXME: vec_push_within_capacity #100486, replace the following with:
+        //
+        // if let Err(value) = self.spans.push_within_capacity(interned) {
+        //     self.spans.try_reserve(1)?;
+        //     // this cannot fail, the previous line either returned or added at least 1 free slot
+        //     let _ = self.spans.push_within_capacity(value);
```
wouldn't that end up calling `try_next_symbol` twice?

With the new code, we try to reserve capacity eagerly:
```
        self.spans.try_reserve(1)?;
        self.spans.push(interned);
```
which unfortunately checks capacity redundantly, the 2nd one being in the push itself, which might impact the performance. The whole point of using `push_within_capacity` is to try to push first, where there is only one capacity check in the optimistic scenario, and only when that fails code branches to allocate.

royaltm commented on src/backend/bucket/mod.rs:
```diff
@@ -142,29 +142,54 @@ where
             let old_head = core::mem::replace(&mut self.head, new_head);
-            self.full.push(old_head.finish());
+            let old_string = old_head.finish();
+            // FIXME: vec_push_within_capacity #100486, replace the following with:
+            //
+            // if let Err(value) = self.full.push_within_capacity(old_string) {
+            //     self.full.try_reserve(1)?;
+            //     // this cannot fail, the previous line either returned or added at least 1 free slot
+            //     let _ = self.full.push_within_capacity(value);
+            // }
```
sure, but that would need to be a generic function that takes a vector as argument because `spans` and `full` are vectors of different types

royaltm commented:
I'm beginning to think that without stabilization of `push_within_capacity` we can't do much to improve backend's fill performance, as we ended up with an algorithm that checks the capacity of the vector multiple times for a single push operation. Please see the comment above https://github.com/Robbepop/string-interner/pull/71#discussion_r1648808767 for more on this.

If I'm not mistaken another area of improvement might be in the raw hash-map department. With the latest changes I have also introduced redundant capacity checks in the `try_get_or_intern_using` function.

royaltm commented:
Also, while I was pondering on the var-int buffer encoding issue if we are going to address the "2 phased encoding step" and we decide to try the single-step array encoding approach I'd also suggest to change the var-int algorithm to the more efficient one:

I've created a gist with more info on the topic and playground examples:
https://gist.github.com/royaltm/f31ce5b1c8878e7c10ff7d335f3826e2

Please feel free to check it out at your leisure.
